pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -397,6 +407,51 @@ func (h *Handler) removeOldNad(setting *harvesterv1.Setting) error {
 	return nil
 }
 
+func (h *Handler) validateIPAddressesAllocations(setting *harvesterv1.Setting) error {
+	if setting.Value == "" {
+		return nil
+	}
+
+	lhnodes, err := h.longhornNodeCache.List(metav1.NamespaceAll, labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list longhorn node cache %v", err)
+	}
+
+	MinAllocatableIPAddrs := 0
+	nodeNum := 1
+	//Required Number of IPs = (Number of Nodes * 2) + (Number of Nodes * Number of Disks * 2) + Number of Images to Download/Upload
+	//Number of Images to download/upload and pod count during upgrade is dynamic, so skipped in the formula calculated.
+	for _, lhNode := range lhnodes {
+		MinAllocatableIPAddrs = MinAllocatableIPAddrs + nodeNum + len(lhNode.Spec.Disks)
+	}
+
+	var config Config
+
+	if err := json.Unmarshal([]byte(setting.Value), &config); err != nil {
+		return fmt.Errorf("parsing value error %v", err)
+	}
+
+	ipprefix := strings.Split(config.Range, "/")
+	ippoolName := ipprefix[0] + "-" + ipprefix[1]
+
+	ippools, err := h.whereaboutsCNIIPPoolCache.List(util.KubeSystemNamespace, labels.Everything())
+	if err != nil {
+		return fmt.Errorf("wherabouts IPPool error %v", err)
+	}
+
+	for _, ippool := range ippools {
+		if ippool.Name != ippoolName {
+			continue
+		}
+
+		if len(ippool.Spec.Allocations) >= MinAllocatableIPAddrs {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("whereabouts cni IP allocation failure for IPPool %s retrying again", ippoolName)
```

Comment:
I think we should make it more clear that there is an error here because `len(ippool.Spec.Allocations) < MinAllocatableIPAddrs`.

Consider something like:

```
return fmt.Errorf("whereabouts cni IP allocation failure for IPPool %s retrying again. 
    required: %d, 
    allocated: %d", 
    ippoolName, 
    MinAllocatableIPAddrs, 
    len(ippool.Spec.Allocations))
```